honeybee_schema/radiance/material.py

```diff
@@ -0,0 +1,314 @@
+        ge=0,
+        le=1,
+        description='A value between 0 and 1 for the blue channel transmissivity '
+                    '(default: 0).'
+    )
+
+    refraction_index: float = Field(
+        default=1.52,
+        ge=0,
+        le=1,
+        description='A value between 0 and 1 for the index of refraction '
+                    '(default: 1.52).'
+    )
+
+
+class BSDF(ModifierBase):
+    """Radiance BSDF (Bidirectional Scattering Distribution Function) material."""
+
+    type: constr(regex='^BSDF$') = 'BSDF'
+
+    up_orientation: List[float] = Field(
+        default=(0.01, 0.01, 1.00),
+        min_items=3,
+        max_items=3,
+        description='Vector as sequence that sets the hemisphere that the BSDF material '
+                    'faces. (default: (0.01, 0.01, 1.00).'
+    )
+
+    thickness: float = Field(
+        default=0,
+        description='Optional number to set the thickness of the BSDF material '
+                    'Sign of thickness indicates whether proxied geometry is '
+                    'behind the BSDF surface (when thickness is positive) or in '
+                    'front (when thickness is negative)(default: 0).'
+    )
+
+    function_file: str = Field(
+        default='.',
+        min_length=1,
+        max_length=100,
+        description='Optional input for function file (default: ".").'
+    )
+
+    transform: str = Field(
+        default=None,
+        min_length=1,
+        max_length=100,
+        description='Optional transform input to scale the thickness and reorient '
+                    'the up vector (default: None).'
+    )
+
+    bsdf_data: bytes = Field(
+        ...,
+        description='BSDF xml file data as bytes.'
+    )
+
+    front_diffuse_reflectance: List[float] = Field(
+        default=None,
+        min_items=3,
+        max_items=3,
+        description='Optional additional front diffuse reflectance as sequence of '
+                    'numbers (default: None).'
+    )
+
+    back_diffuse_reflectance: List[float] = Field(
+        default=None,
+        min_items=3,
+        max_items=3,
+        description='Optional additional back diffuse reflectance as sequence of '
+                    'numbers (default: None).'
+    )
+
+    diffuse_transmittance: List[float] = Field(
+        default=None,
+        min_items=3,
+        max_items=3,
+        description='Optional additional diffuse transmittance as sequence of '
+                    'numbers (default: None).'
+    )
+
+    @staticmethod
+    def _is_valid(arr):
+        """Recurse boolean value check"""
+        valid = True
+        for n in arr:
+            valid = valid and 0 <= n <= 1
+        return valid
+
+    @validator('front_diffuse_reflectance')
+    def check_front_diff_value(cls, values):
+        """Ensure every list value is between 0 and 1."""
+        assert cls._is_valid(values), \
+            'Every value in front diffuse reflectance must be between 0 and 1.'
```

Comment:
I think calling another validator here is probably too much. You can do something like this:

```suggestion
        assert all(0 <=v <= 1 for v in values), \
            'Every value in front diffuse reflectance must be between 0 and 1.'
```

Reply:
Woah! I didn't even know python had an all() function. I'm always frustrated that Python took out the reduce function, but this helps fill that gap.